geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,465 @@
+
+    if (isCancel) {
+      return cancelReplicateRegion(context, regionName, senderId);
+    }
+
+    final Region region = cache.getRegion(regionName);
+    if (region == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.REPLICATE_REGION__MSG__REGION__NOT__FOUND, regionName));
+    }
+
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (sender == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.REPLICATE_REGION__MSG__SENDER__NOT__FOUND, senderId));
+    }
+
+    if (sender.isParallel() && !(region instanceof PartitionedRegion)) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.REPLICATE_REGION__MSG__CANNOT__REPLICATE__NON__PARTITIONED__REGION__WITH__PARALLEL__SENDER);
+    }
+
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.REPLICATE_REGION__MSG__SENDER__NOT__RUNNING, senderId));
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          CliStrings.format(CliStrings.REPLICATE_REGION__MSG__SENDER__SERIAL__AND__NOT__PRIMARY,
+              senderId));
+    }
+
+    try {
+      return executeReplicateFunctionInNewThread(context, region, regionName, sender, maxRate,
+          batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.REPLICATE_REGION__MSG__EXECUTION__CANCELED);
+    } catch (ExecutionException e) {
+      Writer buffer = new StringWriter();
+      PrintWriter pw = new PrintWriter(buffer);
+      e.printStackTrace(pw);
+      logger.error("Exception when running replicate command: {}", buffer.toString());
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.REPLICATE_REGION__MSG__EXECUTION__FAILED, e.getMessage()));
+    }
+  }
+
+  private CliFunctionResult executeReplicateFunctionInNewThread(FunctionContext<Object[]> context,
+      Region region, String regionName, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    Callable<CliFunctionResult> callable =
+        new ReplicateRegionCallable(context, region, sender, maxRate, batchSize);
+    FutureTask<CliFunctionResult> futureTask = new FutureTask<>(callable);
+    ExecutorService executor = LoggingExecutors
+        .newSingleThreadExecutor(getReplicateRegionFunctionThreadName(regionName,
+            sender.getId()), true);
+    executor.execute(futureTask);
```

Comment:
Should ReplicateRegionFunction create an ExecutorService when it is instantiated rather than creating one for every executeReplicateFunctionInNewThread call?

Reply:
The problem is that we need to pass to the ExecutorService the thread name (dependent on the region name and senderId) and that is not known at instantiation time but rather when the execute method is invoked. This is needed to be able to cancel an ongoing replicate command by using a known thread name.
Do you see any problem with this?